File: ios/sensors/GyrStreamManager.swift
import Foundation
import PolarBleSdk
import RxSwift
import React

class GyrStreamManager {

    private weak var bridge: PolarBridge?
    private var api: PolarBleApi?

    private var gyrDisposable: Disposable?
    private var isGyrStreaming = false

    private var gyrBuffer: [[String: Any]] = []
    private let gyrBufferQueue = DispatchQueue(label: "com.polarbridge.gyrBufferQueue")
    private var gyrFlushTimer: Timer?

    private let SENSOR_BUFFER_MS: TimeInterval = 10_000
    private let disposeBag = DisposeBag()

    init(api: PolarBleApi?, bridge: PolarBridge) {
        self.api = api
        self.bridge = bridge
    }

    // MARK: - Public

    func fetchGyrData(_ deviceId: String, bufferMs: NSNumber?) {
        NSLog("GyrStreamManager: Fetch GYR Data called on: \(deviceId) bufferMs: \(String(describing: bufferMs))")

        let resolvedBufferMs: TimeInterval
        if let number = bufferMs, !(number is NSNull) {
            resolvedBufferMs = number.doubleValue >= 0 ? number.doubleValue : SENSOR_BUFFER_MS
        } else {
            resolvedBufferMs = SENSOR_BUFFER_MS
        }

        guard let api = api else {
            NSLog("GyrStreamManager: Polar API not initialized")
            return
        }

        if isGyrStreaming {
            disposeGyrStream()
            stopGyrFlushTimer()
            flushGyrBuffer()
            isGyrStreaming = false

            bridge?.sendEvent(
                withName: PolarEvent.PolarGyrComplete.rawValue,
                body: ["message": "GYR Stream stopped"]
            )
            return
        }

        isGyrStreaming = true
        startGyrFlushTimer(bufferMs: resolvedBufferMs)

        gyrDisposable = SensorSettings.requestStreamSettings(
                api: api,
                identifier: deviceId,
                feature: .gyro
            )
            .flatMap { settings in
                api.startGyroStreaming(deviceId, settings: settings).asObservable()
            }
            .observe(on: MainScheduler.instance)
            .subscribe(
                onNext: { [weak self] gyrData in
                    guard let self = self else { return }

                    self.gyrBufferQueue.async {
                        for sample in gyrData {
                            var event: [String: Any] = [:]
                            event["gyrX"] = "\(sample.x)"
                            event["gyrY"] = "\(sample.y)"
                            event["gyrZ"] = "\(sample.z)"
                            event["gyrTimestamp"] = Double(sample.timeStamp)

                            self.gyrBuffer.append(event)
                        }
                    }
                },
                onError: { [weak self] error in
                    guard let self = self else { return }

                    self.stopGyrFlushTimer()
                    self.flushGyrBuffer()
                    self.isGyrStreaming = false

                    self.bridge?.sendEvent(
                        withName: PolarEvent.PolarGyrError.rawValue,
                        body: ["error": error.localizedDescription]
                    )

                    self.gyrDisposable = nil
                },
                onCompleted: { [weak self] in
                    guard let self = self else { return }

                    self.stopGyrFlushTimer()
                    self.flushGyrBuffer()
                    self.isGyrStreaming = false

                    self.bridge?.sendEvent(
                        withName: PolarEvent.PolarGyrComplete.rawValue,
                        body: ["message": "GYR stream complete"]
                    )

                    self.gyrDisposable = nil
                }
            )

        gyrDisposable?.disposed(by: disposeBag)
    }

    func disposeGyrStream() {
        isGyrStreaming = false
        gyrDisposable?.dispose()
    }

    // MARK: - Buffer

    private func startGyrFlushTimer(bufferMs: TimeInterval) {
        stopGyrFlushTimer()

        let intervalSeconds = bufferMs / 1000.0

        DispatchQueue.main.async {
            self.gyrFlushTimer = Timer(
                timeInterval: intervalSeconds,
                repeats: true
            ) { [weak self] _ in
                self?.flushGyrBuffer()
            }

            RunLoop.main.add(self.gyrFlushTimer!, forMode: .common)
        }
    }

    private func stopGyrFlushTimer() {
        DispatchQueue.main.async {
            self.gyrFlushTimer?.invalidate()
            self.gyrFlushTimer = nil
        }
    }

    private func flushGyrBuffer() {
        gyrBufferQueue.async {
            guard !self.gyrBuffer.isEmpty else { return }

            let events = self.gyrBuffer
            self.gyrBuffer.removeAll()

            DispatchQueue.main.async {
                for event in events {
                    self.bridge?.sendEvent(
                        withName: PolarEvent.PolarGyrData.rawValue,
                        body: event
                    )
                }
            }
        }
    }
}

File: android/src/main/java/com/polarbridge/sensors/GyrStreamManager.kt
package com.polarbridge.sensors

import android.util.Log
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap
import com.polar.sdk.api.PolarBleApi
import com.polar.sdk.api.errors.PolarInvalidArgument
import com.polar.sdk.api.model.PolarSensorSetting
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import java.util.concurrent.TimeUnit

class GyrStreamManager(
  private val api: PolarBleApi,
  private val sendEvent: (String, WritableMap) -> Unit,
  private val TAG: String = "GyrStreamManager"
) {
  private var gyrDisposable: Disposable? = null

  fun fetchGyrData(deviceId: String, sensorSettings: SensorSettings, ms: Double?, sensorBufferMs: Long) {
    Log.d(TAG, "Fetch Gyroscope Data called on: $deviceId ")
    val bufferMs = ms?.toLong()?.takeIf { it >= 0 } ?: sensorBufferMs
    val isDisposed = gyrDisposable?.isDisposed ?: true
    try {
      if (isDisposed) {
        gyrDisposable = sensorSettings.requestStreamSettings(deviceId, PolarBleApi.PolarDeviceDataType.GYRO)
          .flatMap { settings: PolarSensorSetting ->
            api.startGyroStreaming(deviceId, settings)
          }
          .flatMapIterable { it.samples }
          .buffer(bufferMs, TimeUnit.MILLISECONDS)
          .observeOn(AndroidSchedulers.mainThread())
          .subscribe(
            { samples ->
              if (samples.isEmpty()) return@subscribe

              Log.d(TAG, "Flushing GYR buffer (${samples.size} samples)")

              for (data in samples) {
                // Log.d(TAG, "GYR    x: ${data.x} y: ${data.y} z: ${data.z} timeStamp: ${data.timeStamp}")

                val event: WritableMap = Arguments.createMap()
                // Float not supported
                // See: https://github.com/facebook/react-native/issues/9685
                // See: https://javadoc.io/doc/com.facebook.react/react-native/0.20.1/com/facebook/react/bridge/WritableMap.html
                event.putString("gyrX", "${data.x}")
                event.putString("gyrY", "${data.y}")
                event.putString("gyrZ", "${data.z}")
                // Long not supported, use double as workaround
                event.putDouble("gyrTimestamp", data.timeStamp.toDouble())

                sendEvent("PolarGyrData", event)
              }
            },
            { error: Throwable ->
              Log.e(TAG, "GYR stream failed. Reason $error")

              val errorEvent = Arguments.createMap()
              errorEvent.putString("error", error.toString())
              sendEvent("PolarGyrError", errorEvent)
            },
            {
              Log.d(TAG, "GYR stream complete")

              val completeEvent = Arguments.createMap()
              completeEvent.putString("message", "GYR stream complete")
              sendEvent("PolarGyrComplete", completeEvent)
            }
          )
      } else {
        disposeGyrStream()
        Log.d(TAG, "GYR stream stopped")
      }
    } catch(polarInvalidArgument: PolarInvalidArgument){
      Log.e(TAG, "Failed to fetch GYR Data. Reason $polarInvalidArgument ")
    }
  }

  fun disposeGyrStream(){
    try {
      gyrDisposable?.dispose()
    } catch (e: Exception) {
      Log.w(TAG, "Failed to dispose GYR stream cleanly. Bluetooth service likely died.", e)
    } finally {
      gyrDisposable = null
    }
  }
}
